edx-solutions/xblock-drag-and-drop-v2
# Add option for limiting number of attempts in assessment mode to Studio editor

**Description**

This PR implements the following use case<sup>1</sup>:

> As a course author, I want to be able to limit the number of times a learner may attempt a drag-and-drop problem in assessment mode. More specifically, I want to be able to specify that a learner may try a problem one time, n times, or as many times as they wish (unlimited attempts).

<sup>1</sup>Note that learner-facing functionality related to limiting the number of available attempts will be covered by future PRs.

*UX*

For drag-and-drop problems in standard mode, the interactive Studio editor does not show an option for setting the maximum number of attempts. 

For drag-and-drop problems in assessment mode, the interactive Studio editor shows a numeric input field that allows course authors to specify the maximum number of attempts. The field is hidden until a course author sets the mode of a drag-and-drop problem to "assessment mode" (as described above).

* By leaving the maximum number of attempts unset, course authors can indicate that the number of times learners may attempt the problem is unlimited. 
* By setting the maximum number of attempts to a positive number `n`, course authors can indicate that learners may attempt the problem exactly `n` times. 

The value of the “Maximum attempts” setting defaults to “not set” (unlimited attempts).

Switching modes does not affect the current value of the “Maximum attempts” field.

**JIRA tickets**: [SOL-1941](https://openedx.atlassian.net/browse/SOL-1941)

**Discussions**: The feature implemented in this PR was discussed with @marcotuts as part of his review of the [discovery document](https://docs.google.com/document/d/1rFliZo1vlohJ7cde7KfP5bgEk0QqcvT1VNnUUa9_vjE/edit) for DnDv2 assessment mode.

* Relevant section: 2.1.2
* Relevant comment threads (have since been "marked as resolved"):
  * https://docs.google.com/document/d/1rFliZo1vlohJ7cde7KfP5bgEk0QqcvT1VNnUUa9_vjE/edit?disco=AAAAAsmOs3E
  * https://docs.google.com/document/d/1rFliZo1vlohJ7cde7KfP5bgEk0QqcvT1VNnUUa9_vjE/edit?disco=AAAAAsmOs3I

**Dependencies**: None

**Screenshots**:

Standard mode:

![edit-view-standard](https://cloud.githubusercontent.com/assets/961441/16955907/a3491718-4dd6-11e6-99a1-edc9033feb9e.png)

Assessment mode:

![edit-view](https://cloud.githubusercontent.com/assets/961441/16955744/20e10984-4dd6-11e6-82af-6a950401d362.png)

**Sandbox URL**:

Studio: http://studio-dndv2-sandbox2.opencraft.hosting/
LMS: http://dndv2-sandbox2.opencraft.hosting/

The sandbox is currently using version https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/81/commits/40037b19bcf08ec18f95aa9aa33dc0d1e96c1531 of DnDv2.

**Testing instructions**:

*To test from scratch* (using sandbox or local development environment):

1. Add a DnDv2 block to a unit in Studio.
2. Click "EDIT". Observe that "Problem mode" is set to "standard" (cf. #77), and that the "Maximum attempts" setting is not visible.
3. Change "Problem mode" to "assessment". Observe that the "Maximum attempts" setting becomes visible below the "Problem mode" setting.
4. (Optionally) Change the value of the "Maximum attempts setting.
5. Continue to the last pane of the Studio editor, then click "Save".
6. Click "EDIT" again. Observe that the "Maximum attempts setting is still visible (and set to whatever value it had when you clicked "Save").
7. Change "Problem mode" to "standard". Observe that the "Maximum attempts" setting is hidden.
8. Save, click "EDIT" again, and switch "Problem mode" back to "assessment". Observe that the value of the "Maximum attempts" has been preserved from when you last changed it (or is still at the default if you chose not to change it earlier).

*To test using existing DnDv2 instances*, access

"PR Review Testing" > "Maximum attempts" > "Standard mode"

and/or

"PR Review Testing" > "Maximum attempts" > "Assessment mode"

units on the sandbox (they are part of the "Drag and Drop Demos" course) and perform relevant steps listed above.

**Reviewers**

- [x] OpenCraft/Solutions: @mtyaka 
- [ ] a11y:  @cptvitamin and/or @clrux

Also tagging @pdesjardins as a docs FYI.
